alf/utils/losses.py

```diff
@@ -714,54 +714,48 @@ class BipartiteMatchingLoss(object):
         """
         super().__init__()
-        self._pair_loss_fn = pair_loss_fn
         self._reduction = reduction
         assert reduction in ['mean', 'sum', 'none']
         self._name = name
 
     def forward(self,
-                prediction: Tensor,
-                target: Tensor,
-                target_mask: Tensor = None):
-        """
+                matching_cost_mat: torch.Tensor,
+                cost_mat: torch.Tensor = None):
+        """Compute the optimal matching loss.
+
         Args:
-            prediction: the predicted set with a shape of ``[B,N,...]``.
-            target: the target set with a shape of ``[B,N,...]``.
-            target_mask: the valid mask for the target set. We assume that the
-                target and prediction sets each always contains ``N`` objects, but
-                some objects in the target set are just paddings whose mask values
-                are 0. These paddings will always have a matching loss of 0. The
-                shape should be ``[B,N]``. If None, then all target objects are
-                valid.
+            matching_cost_mat: the cost matrix used to determine the optimal
+                matching. It shape should be ``[B,N,N]``.
+            cost_mat: the cost matrix used to compute the optimal loss once the
+                optimal matching is found. According to the DETR paper, this
+                cost matrix might be different from the one used for matching.
+                If None, then it will be the same matrix for matching.
+
+        Returns:
+            tuple:
+            - the optimal loss. If reduction is 'mean' or 'sum', its shape is
+              ``[B,N]``, otherwise its shape is ``[B,N,N]``.
+            - the optimal matching given the cost matrix. Its shape is ``[B,N]``,
+              where the value of n-th entry is its mapped index in the target set.
         """
-        assert prediction.shape[:2] == target.shape[:2]
-        B, N = prediction.shape[:2]
-        cost_mat = self._pair_loss_fn(prediction, target)  # [B,N,N]
-        assert cost_mat.shape == (B, N, N), (
-            "The pairwise loss function must enumerate all pairs and output "
-            "a scalar loss for each pair!")
-
-        # mask out any cost with mask values=0
-        if target_mask is not None:
-            target_mask = target_mask.unsqueeze(1)  # [B,1,N]
-            cost_mat = cost_mat * target_mask
+        if cost_mat is None:
+            cost_mat = matching_cost_mat
 
         with torch.no_grad():
+            B, N = matching_cost_mat.shape[:2]
             # [B*N, B*N]
-            max_cost = cost_mat.max() + 1.
-            big_cost_mat = torch.block_diag(*list(cost_mat - max_cost))
+            max_cost = matching_cost_mat.max() + 1.
+            big_cost_mat = torch.block_diag(
+                *list(matching_cost_mat - max_cost))
             # fill in all off-diag entries with a max cost
             big_cost_mat = big_cost_mat + max_cost
```

Comment:
This will also change the diagonal elements?

Reply:
> This will also change the diagonal elements?

The diagonal elements have been subtracted by `max_cost`.

Reply:
> The comment is a little bit confusing. Adding the same value to all the elements is unnecessary? It will not change the solution.

The max_cost is expected to only affect the off-diagonal entries. Ideally, for each sample b of B, we need to call linear_sum_assignment on the NxN matrix to find a matching. But this is slow. So I created a big matrix of size BNxBN. In order to make sure different NxN matrices won't affect each other, we need to make sure the off-diagonal values are very large, so that matchings are only found within each diagonal block.

Reply:
I understand your motivation. But I think you don't need to add max_cost. The off-diagonal elements(i,e,zeros) are already guaranteed to be bigger than the diagonal elements.

Reply:
> I understand your motivation. But I think you don't need to add max_cost. The off-diagonal elements(i,e,zeros) are already guaranteed to be bigger than the diagonal elements.

I see. yeah there is no need to add again.